src/block/Beacon.php

```diff
@@ -23,11 +23,209 @@
+			}
+		}
+	}
+
+	public function getBeaconLevel() : int {
+		$beaconLevel = 0;
+		for($i = self::MIN_LEVEL_BEACON; $i <= self::MAX_LEVEL_BEACON; $i++){
+			if(!$this->isBeaconLevelValid($i)){
+				break;
+			}
+			$beaconLevel++;
+		}
+		return $beaconLevel;
+	}
+
+	public function getPrimaryEffect() : ?Effect{
+		return $this->primaryEffect;
+	}
+
+	/** @return $this */
+	public function setPrimaryEffect(?Effect $primaryEffect) : self{
+		$this->primaryEffect = $primaryEffect;
+		return $this;
+	}
+
+	public function getSecondaryEffect() : ?Effect{
+		return $this->secondaryEffect;
+	}
+
+	/** @return $this */
+	public function setSecondaryEffect(?Effect $secondaryEffect) : self{
+		$this->secondaryEffect = $secondaryEffect;
+		return $this;
+	}
+
 	public function getLightLevel() : int{
 		return 15;
 	}
 
-	//TODO
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($player instanceof Player){
+			$player->setCurrentWindow(new BeaconInventory($this->position));
+		}
+
+		return true;
+	}
+
+	public function onScheduledUpdate() : void{
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 20 * 3);
+
+		if($this->primaryEffect === null){
+			return;
+		}
+		if(($beaconLevel = $this->getBeaconLevel()) >= self::MIN_LEVEL_BEACON){
+			if(!$this->viewSky()){
+				return;
+			}
+
+			$radius = (10 * $beaconLevel) + 10;
+			$effectDuration = 9 + (2 * $beaconLevel);
+
+			$world = $this->position->getWorld();
+			$aabb = $this->getCollisionBoxes()[0]->expandedCopy($radius, $radius, $radius)->addCoord(0, $world->getMaxY(), 0);
+			if($this->primaryEffect === $this->secondaryEffect){
+				foreach($world->getNearbyEntities($aabb) as $entity){
+					if($entity instanceof Player){
+						$entity->getEffects()->add(new EffectInstance($this->primaryEffect, $effectDuration * 20, 1));
+					}
+				}
+			}else{
+				$effects = [$this->primaryEffect];
+				if($this->secondaryEffect !== null){
+					$effects[] = $this->secondaryEffect;
+				}
+				foreach($world->getNearbyEntities($aabb) as $entity){
+					if($entity instanceof Player){
+						foreach($effects as $effect){
+							$entity->getEffects()->add(new EffectInstance($effect, $effectDuration * 20, 0));
+						}
+					}
+				}
+			}
+		}
+	}
+
+	public function isBeaconLevelValid(int $level) : bool{
+		if($level < self::MIN_LEVEL_BEACON || $level > self::MAX_LEVEL_BEACON){
+			throw new InvalidArgumentException("Beacon level must be in range " . self::MIN_LEVEL_BEACON . "-" . self::MAX_LEVEL_BEACON . ", $level given");
+		}
+
+		$world = $this->position->getWorld();
+		$pos = $this->position->subtract(0, $level, 0);
+		for($x = -$level; $x <= $level; $x++){
+			for($z = -$level; $z <= $level; $z++){
+				$block = $world->getBlock($pos->add($x, 0, $z));
+				if(!isset(self::ALLOWED_BLOCK_IDS[$block->getTypeId()])){
+					return false;
```

Comment:
I really don't like this. I know that's the most obvious way to do it, but it's going to lag like shit.